File: src/tools.rs
use crate::{map::Warehouse, model::*, planner::astar, Simulation};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::BTreeSet;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case", deny_unknown_fields)]
pub enum Command {
    GetState,
    GetKpis,
    GetEvents {
        #[serde(default)]
        after: u64,
    },
    Reset {
        map: Warehouse,
        robots: Vec<Position>,
        seed: u64,
    },
    AddOrder {
        pickup: Position,
        dropoff: Position,
        #[serde(default)]
        priority: u8,
    },
    GenerateOrders {
        count: usize,
    },
    SetStrategy {
        strategy: Strategy,
    },
    Dispatch,
    AssignOrder {
        robot_id: u32,
        order_id: u32,
    },
    PlanPath {
        start: Position,
        goal: Position,
        #[serde(default)]
        avoid: BTreeSet<Position>,
    },
    Replan {
        robot_id: u32,
        #[serde(default)]
        avoid: BTreeSet<Position>,
    },
    SetBlocked {
        position: Position,
        blocked: bool,
    },
    InjectFault {
        robot_id: u32,
    },
    RepairRobot {
        robot_id: u32,
    },
    Step {
        ticks: u32,
    },
}

pub fn execute(sim: &mut Simulation, command: &Command) -> Result<Value, String> {
    match command {
        Command::GetState => Ok(json!({ "world": sim, "kpis": sim.kpis() })),
        Command::GetKpis => Ok(json!(sim.kpis())),
        Command::GetEvents { after } => Ok(json!(sim
            .events
            .iter()
            .filter(|e| e.sequence > *after)
            .collect::<Vec<_>>())),
        Command::Reset { map, robots, seed } => {
            *sim = Simulation::new(map.clone(), robots.clone(), *seed)?;
            Ok(json!({"reset": true}))
        }
        Command::AddOrder {
            pickup,
            dropoff,
            priority,
        } => Ok(json!({ "order_id": sim.add_order(*pickup, *dropoff, *priority)? })),
        Command::GenerateOrders { count } => {
            Ok(json!({ "order_ids": sim.generate_orders(*count)? }))
        }
        Command::SetStrategy { strategy } => {
            sim.strategy = *strategy;
            Ok(json!({ "strategy": strategy }))
        }
        Command::Dispatch => Ok(json!({ "assigned": sim.dispatch() })),
        Command::AssignOrder { robot_id, order_id } => {
            sim.assign(*robot_id, *order_id)?;
            Ok(json!({ "assigned": true }))
        }
        Command::PlanPath { start, goal, avoid } => {
            if avoid.iter().any(|p| !sim.map.contains(*p)) {
                return Err("avoid cell outside map".into());
            }
            let path = astar(&sim.map, *start, *goal, avoid).ok_or("no path")?;
            Ok(json!({ "distance": path.len(), "path": path }))
        }
        Command::Replan { robot_id, avoid } => {
            Ok(json!({ "distance": sim.replan(*robot_id, avoid)? }))
        }
        Command::SetBlocked { position, blocked } => {
            sim.set_blocked(*position, *blocked)?;
            Ok(json!({ "blocked": blocked }))
        }
        Command::InjectFault { robot_id } => {
            sim.inject_fault(*robot_id)?;
            Ok(json!({ "faulted": true }))
        }
        Command::RepairRobot { robot_id } => {
            sim.repair(*robot_id)?;
            Ok(json!({ "repaired": true }))
        }
        Command::Step { ticks } => {
            sim.step(*ticks)?;
            Ok(json!(sim.kpis()))
        }
    }
}

pub fn response(sim: &mut Simulation, command: &Command) -> Value {
    match execute(sim, command) {
        Ok(data) => json!({ "ok": true, "tick": sim.tick, "data": data }),
        Err(message) => {
            json!({ "ok": false, "tick": sim.tick, "error": { "code": "invalid_operation", "message": message } })
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Replay {
    pub version: u32,
    pub seed: u64,
    pub commands: Vec<Command>,
}
impl Replay {
    pub fn run(&self) -> Result<Simulation, String> {
        if self.version != 1 {
            return Err("unsupported replay version".into());
        }
        let mut sim = Simulation::demo(self.seed);
        // Accepted, typed commands are recorded, including rejected operations.
        // Their deterministic errors do not prevent subsequent commands replaying.
        for command in &self.commands {
            let _ = execute(&mut sim, command);
        }
        Ok(sim)
    }
}

File: agent/tools.mjs
const position = {
  type: 'object',
  description: 'Grid position. x increases to the right, y increases downward.',
  properties: {
    x: { type: 'integer', minimum: 0, maximum: 255 },
    y: { type: 'integer', minimum: 0, maximum: 255 },
  },
  required: ['x', 'y'],
  additionalProperties: false,
};

const positionList = {
  type: 'array',
  items: position,
  uniqueItems: true,
};

const robotId = {
  type: 'integer',
  minimum: 1,
  maximum: 4294967295,
};

const orderId = {
  type: 'integer',
  minimum: 1,
  maximum: 4294967295,
};

const map = {
  type: 'object',
  properties: {
    width: { type: 'integer', minimum: 1, maximum: 256 },
    height: { type: 'integer', minimum: 1, maximum: 256 },
    obstacles: positionList,
    blocked: positionList,
  },
  required: ['width', 'height', 'obstacles'],
  additionalProperties: false,
};

function tool(name, description, properties = {}, required = []) {
  return {
    type: 'function',
    function: {
      name,
      description,
      parameters: {
        type: 'object',
        properties,
        required,
        additionalProperties: false,
      },
    },
  };
}

export const SIM_TOOLS = [
  tool('get_state', 'Read the complete simulation state: map, robots, orders, tick, strategy, events and heatmap.'),
  tool('get_kpis', 'Read current throughput, completion time, utilization, distance and wait metrics.'),
  tool('get_events', 'Read new domain events after an event sequence number.', {
    after: { type: 'integer', minimum: 0 },
  }),
  tool('reset', 'Replace the world with a new map and robot set. This resets tick and metrics.', {
    map,
    robots: { ...positionList, minItems: 1, maxItems: 64 },
    seed: { type: 'integer', minimum: 0 },
  }, ['map', 'robots', 'seed']),
  tool('add_order', 'Add one order with a pickup cell, dropoff cell and optional priority.', {
    pickup: position,
    dropoff: position,
    priority: { type: 'integer', minimum: 0, maximum: 9 },
  }, ['pickup', 'dropoff']),
  tool('generate_orders', 'Generate random connected orders using the fixed simulator seed.', {
    count: { type: 'integer', minimum: 1, maximum: 1000 },
  }, ['count']),
  tool('set_strategy', 'Set baseline dispatch strategy. Use manual when the agent owns decisions.', {
    strategy: { type: 'string', enum: ['nearest', 'balanced', 'manual'] },
  }, ['strategy']),
  tool('dispatch', 'Immediately assign pending orders using the current baseline strategy.'),
  tool('assign_order', 'Assign a pending order to an idle robot.', {
    robot_id: robotId,
    order_id: orderId,
  }, ['robot_id', 'order_id']),
  tool('plan_path', 'Plan a shortest path from start to goal, optionally avoiding extra cells.', {
    start: position,
    goal: position,
    avoid: positionList,
  }, ['start', 'goal']),
  tool('replan', 'Replan the current path for a robot that already has an active order.', {
    robot_id: robotId,
    avoid: positionList,
  }, ['robot_id']),
  tool('set_blocked', 'Block or unblock a floor cell. Cannot block a cell currently occupied by a robot.', {
    position,
    blocked: { type: 'boolean' },
  }, ['position', 'blocked']),
  tool('inject_fault', 'Force a robot into the faulted state. Its unfinished order returns to pending.', {
    robot_id: robotId,
  }, ['robot_id']),
  tool('repair_robot', 'Repair a faulted robot and make it idle again.', {
    robot_id: robotId,
  }, ['robot_id']),
  tool('step', 'Advance the simulation by the given number of ticks. Only step advances time.', {
    ticks: { type: 'integer', minimum: 1, maximum: 10000 },
  }, ['ticks']),
];

const names = new Set(SIM_TOOLS.map((entry) => entry.function.name));

export async function executeTool(client, name, args = {}) {
  if (!names.has(name)) {
    return { ok: false, error: `Unknown tool: ${name}` };
  }
  try {
    const data = await client.call(name, args);
    return { ok: true, data };
  } catch (error) {
    return { ok: false, error: error.message };
  }
}
